exchanges/btcc/btcc_websocket.go

```diff
@@ -559,3 +469,62 @@ func (b *BTCC) WsProcessOldOrderbookSnapshot(ob WsOrderbookSnapshotOld, symbol s
 
 	return nil
 }
+
+// GenerateDefaultSubscriptions Adds default subscriptions to websocket to be handled by ManageSubscriptions()
+func (b *BTCC) GenerateDefaultSubscriptions() {
+	b.Websocket.ChannelsToSubscribe = append(b.Websocket.ChannelsToSubscribe, exchange.WebsocketChannelSubscription{
+		Channel: "SubscribeAllTickers",
+	})
+
+	var channels = []string{"SubOrderBook", "GetTrades", "Subscribe"}
+	enabledCurrencies := b.GetEnabledCurrencies()
+	for i := range channels {
+		for j := range enabledCurrencies {
+			params := make(map[string]interface{})
+			if channels[i] == "SubOrderBook" {
+				params["len"] = "100"
+			} else if channels[i] == "GetTrades" {
+				params["count"] = "100"
+			}
+			b.Websocket.ChannelsToSubscribe = append(b.Websocket.ChannelsToSubscribe, exchange.WebsocketChannelSubscription{
+				Channel:  channels[i],
+				Currency: enabledCurrencies[j],
+				Params:   params,
+			})
+		}
+	}
+}
+
+// Subscribe sends a websocket message to receive data from the channel
+func (b *BTCC) Subscribe(channelToSubscribe exchange.WebsocketChannelSubscription) error {
+	subscription := WsOutgoing{
+		Action: channelToSubscribe.Channel,
+		Symbol: channelToSubscribe.Currency.String(),
+	}
+	if subscription.Action == "SubOrderBook" {
+		subscription.Len = 100
+	} else if subscription.Action == "GetTrades" {
+		subscription.Count = 100
+	}
+
+	time.Sleep(btccWebsocketRateLimit)
+	return b.Conn.WriteJSON(subscription)
```

Comment:
Since we're doing this big overhaul, all websocket writes should be protected by a mutex. Any concurrent websocket writes will cause a panic.

Reply:
Yeah I only encountered that on okex and kraken, but should have expanded it everywhere. Will do 👍